hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,195 @@
+
+#ifndef MPCD_SPHERE_GEOMETRY_H_
+#define MPCD_SPHERE_GEOMETRY_H_
+
+#include <math.h>
+
+#include "BoundaryCondition.h"
+
+#include "hoomd/HOOMDMath.h"
+#include "hoomd/BoxDim.h"
+
+#ifdef NVCC
+#define HOSTDEVICE __host__ __device__ inline
+#else
+#define HOSTDEVICE inline __attribute__((always_inline))
+#include <string>
+#endif // NVCC
+
+namespace mpcd
+{
+namespace detail
+{
+//! Spherical droplet geometry
+/*!
+ * TODO: write proper description of what is being done in the class below
+ */
+class __attribute__((visibility("default"))) SphereGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param R confinement radius
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SlitGeometry(Scalar R, boundary bc)
+            : m_R(R), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the box, and try to avoid branching or absolute value calls. The sign used
+             * in calculations is 1 if the particle is out-of-bounds in the radial direction, 0 otherwise.
+             *
+             * We intentionally use > / < rather than >= / <= to make sure that spurious collisions do not get detected
+             * when a particle is reset to the boundary location. A particle landing exactly on the boundary from the bulk
+             * can be immediately reflected on the next streaming step, and so the motion is essentially equivalent up to
+             * an epsilon of difference in the channel width.
+             */
+            double r2 = pos.x*pos.x + pos.y*pos.y + pos.z*pos.z;
+            const char sign = (r2 > R*R);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*x_dot + y*y_dot + z*z_dot)/r(t)
+            */
+            float vr = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/sqrt(r2);
+            if (sign == 0 || vr == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
```

Comment:
Square roots are expensive, so don't do them if you don't need to. There also isn't a need to create the `sign` variable on line 64, since you aren't doing C math tricks with it. I would just use the logical statement for clarity.
```suggestion
            Scalar vr = dot(pos,vel)
            if (r2 > R*R || vr == Scalar(0))
               {
               dt = Scalar(0);
               return false;
               }
```

Reply:
Instead of r2 > R*R, we need to use r2 < R*R to get TRUE if particle is inside the sphere.

Reply:
Yes, indeed, that inequality was not correct. We need to be careful if we use < or <=. I think we want the second one, but need to check the SlitGeometry for consistency.

Reply:
In the SlitGeometry file, the inequality used is <. I will keep the same for now and we can think about it soon after the second review.